model/dyn_core.F90

```diff
@@ -2252,61 +2242,61 @@ subroutine diff3d(npx, npy, npz, nq, ua, va, w, q,       &
 ! KGao: make ut and vt private as suggested by Lucas
 
 !$OMP parallel do default(none) shared(npz,is,ie,js,je,ua,va,w,dx,dy,rarea, &
-!$OMP                           dudx,dudy,dvdx,dvdy,dwdx,dwdy)              &
+!$OMP                           dudx,dudy,dvdx,dvdy,dwdx,dwdy) &
 !$OMP                           private(ut,vt)
    do k=1,npz
 !-------------------------------------
 ! get du/dy and dv/dx
-       do j=js,je+1
+       do j=js-1,je+1
           do i=is,ie
              vt(i,j) = ua(i,j,k)*dx(i,j)
           enddo
        enddo
        do j=js,je
-          do i=is,ie+1
+          do i=is-1,ie+1
              ut(i,j) = va(i,j,k)*dy(i,j)
           enddo
        enddo
        do j=js,je
           do i=is,ie
-             dudy(i,j,k) = rarea(i,j)*(vt(i,j+1)-vt(i,j))
-             dvdx(i,j,k) = rarea(i,j)*(ut(i+1,j)-ut(i,j))
+             dudy(i,j,k) = 0.5*rarea(i,j)*(vt(i,j+1)-vt(i,j-1))
+             dvdx(i,j,k) = 0.5*rarea(i,j)*(ut(i+1,j)-ut(i-1,j))
           enddo
        enddo
 !-------------------------------------
 ! get du/dx and dv/dy
        do j=js,je
-          do i=is,ie+1
+          do i=is-1,ie+1
              ut(i,j) = ua(i,j,k)*dy(i,j)
           enddo
        enddo
-       do j=js,je+1
+       do j=js-1,je+1
           do i=is,ie
              vt(i,j) = va(i,j,k)*dx(i,j)
           enddo
        enddo
        do j=js,je
           do i=is,ie
-             dudx(i,j,k) = rarea(i,j)*(ut(i+1,j)-ut(i,j))
-             dvdy(i,j,k) = rarea(i,j)*(vt(i,j+1)-vt(i,j))
+             dudx(i,j,k) = 0.5*rarea(i,j)*(ut(i+1,j)-ut(i-1,j))
+             dvdy(i,j,k) = 0.5*rarea(i,j)*(vt(i,j+1)-vt(i,j-1))
           enddo
        enddo
 !-------------------------------------
 ! get dw/dx and dw/dy
        do j=js,je
-          do i=is,ie+1
+          do i=is-1,ie+1
              ut(i,j) = w(i,j,k)*dy(i,j)
           enddo
        enddo
-       do j=js,je+1
+       do j=js-1,je+1
           do i=is,ie
              vt(i,j) = w(i,j,k)*dx(i,j)
           enddo
        enddo
        do j=js,je
           do i=is,ie
-             dwdx(i,j,k) = rarea(i,j)*(ut(i+1,j)-ut(i,j))
-             dwdy(i,j,k) = rarea(i,j)*(vt(i,j+1)-vt(i,j))
+             dwdx(i,j,k) = 0.5*rarea(i,j)*(ut(i+1,j)-ut(i-1,j))
```

Comment:
Please check closely where `dwdx`, etc. are computed and whether a full-grid-length centered difference is needed here. This will compute `dwdx` at cell centers instead of cell interfaces, and necessitates the additional interpolation of `dwdx` below on lines 2338--2357.